src/mono/wasm/runtime/cs-to-js.ts

```diff
@@ -0,0 +1,333 @@
+        }
+    } finally {
+        elemRoot.release();
+    }
+
+    return res;
+}
+
+export function _wrap_delegate_root_as_function(root: WasmRoot) {
+    if (root.value === 0)
+        return null;
+
+    // get strong reference to the Delegate
+    const gc_handle = corebindings._get_js_owned_object_gc_handle(root.value);
+    return _wrap_delegate_gc_handle_as_function(gc_handle);
+}
+
+export function _wrap_delegate_gc_handle_as_function(gc_handle: GCHandle, after_listener_callback?: () => void) {
+    // see if we have js owned instance for this gc_handle already
+    var result = _lookup_js_owned_object(gc_handle);
+
+    // If the function for this gc_handle was already collected (or was never created)
+    if (!result) {
+        // note that we do not implement function/delegate roundtrip
+        result = function () {
+            const delegateRoot = mono_wasm_new_root(get_js_owned_object_by_gc_handle(gc_handle));
+            try {
+                const res = call_method(result[delegate_invoke_symbol], delegateRoot.value, result[delegate_invoke_signature_symbol], arguments);
+                if (after_listener_callback) {
+                    after_listener_callback();
+                }
+                return res;
+            } finally {
+                delegateRoot.release();
+            }
+        };
+
+        // bind the method
+        const delegateRoot = mono_wasm_new_root(get_js_owned_object_by_gc_handle(gc_handle));
+        try {
+            if (typeof result[delegate_invoke_symbol] === "undefined") {
+                result[delegate_invoke_symbol] = cwraps.mono_wasm_get_delegate_invoke(delegateRoot.value);
+                if (!result[delegate_invoke_symbol]) {
+                    throw new Error("System.Delegate Invoke method can not be resolved.");
+                }
+            }
+
+            if (typeof result[delegate_invoke_signature_symbol] === "undefined") {
+                result[delegate_invoke_signature_symbol] = mono_method_get_call_signature(result[delegate_invoke_symbol], delegateRoot.value);
+            }
+        } finally {
+            delegateRoot.release();
+        }
+
+        // NOTE: this would be leaking C# objects when the browser doesn't support FinalizationRegistry. Except in case of EventListener where we cleanup after unregistration.
+        if (_use_finalization_registry) {
+            // register for GC of the deleate after the JS side is done with the function
+            _js_owned_object_registry.register(result, gc_handle);
+        }
+
+        // register for instance reuse
+        // NOTE: this would be leaking C# objects when the browser doesn't support FinalizationRegistry/WeakRef. Except in case of EventListener where we cleanup after unregistration.
+        _register_js_owned_object(gc_handle, result);
+    }
+
+    return result;
+}
+
+export function mono_wasm_create_cs_owned_object(core_name: CharPtr, args: VoidPtr, is_exception: Int32Ptr) {
```

Comment:
core_name is a MonoString and args is a MonoObject. It's not valid to put bare pointers into GC root slots, afaik